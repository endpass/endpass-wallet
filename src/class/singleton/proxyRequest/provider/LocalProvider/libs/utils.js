import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
