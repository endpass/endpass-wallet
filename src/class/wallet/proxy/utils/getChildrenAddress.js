export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
