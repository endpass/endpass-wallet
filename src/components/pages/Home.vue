<template>
  <div class="home-page app-page">
    <div
      v-if="address"
      class="auth-content"
    >
      <div
        class="section section-address"
        data-test="address-card"
      >
        <div class="container">
          <div class="card">
            <div class="card-header">
              <p class="card-header-title">
                Your Address
              </p>
            </div>
            <div class="card-content">
              <div class="columns">
                <div class="column">
                  <account :address="address" />
                </div>
                <div
                  v-if="isExportable"
                  :class="{ 'has-text-centered': !isFaucet }"
                  class="column"
                >
                  <router-link
                    :to="{ name: 'ExportWallet' }"
                    class="button is-warning"
                    data-test="export-wallet-button"
                  >
                    Export Private Key
                  </router-link>
                </div>
                <div
                  v-if="isFaucet"
                  class="column"
                >
                  <v-faucet-button
                    :address="address"
                    :disabled="isFaucetDisable"
                    class="button is-warning"
                    data-test="get-test-eth-button"
                    @donate="onDonate"
                    @donate-error="onDonateError"
                  >
                    {{ faucetTitle }}
                  </v-faucet-button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>

      <div
        v-if="currentNetUserTokensList.length > 0"
        class="section section-tokens"
      >
        <div class="container">
          <div class="card">
            <div class="card-header">
              <p class="card-header-title">
                Your Tokens
              </p>
              <div class="card-header-icon">
                <router-link
                  :to="{ name: 'TokensPage' }"
                  class="button is-outlined is-info is-small"
                  data-test="edit-tokens-button"
                >
                  Edit
                </router-link>
              </div>
            </div>
            <div class="card-content">
              <tokens-list :tokens="currentNetUserTokensList" />
            </div>
          </div>
        </div>
      </div>

      <div class="section section-tokens">
        <div class="container">
          <div class="card">
            <div class="card-header">
              <p class="card-header-title">
                Current Account Tokens ({{ address }})
              </p>
            </div>
            <div class="card-content">
              <tokens-list :tokens="currentAccountTokensList" />
            </div>
          </div>
        </div>
      </div>
    </div>

    <div v-else>
      <div class="section">
        <div class="has-text-centered">
          <div class="card app-card main-app-card">
            <div class="card-content">
              <div v-if="isLoggedIn">
                <h1 class="title">
                  Welcome
                </h1>
                <p class="subtitle">
                  Get started by generating an Ethereum wallet.
                </p>
                <div class="is-centered">
                  <router-link
                    :to="{ name: 'NewWallet', query: $route.query }"
                    class="button is-success is-cta"
                  >
                    Create New Wallet
                  </router-link>
                </div>
              </div>
              <div v-else>
                <p class="subtitle">
                  Please log in to continue.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState } from 'vuex';
import VueTimers from 'vue-timers/mixin';
import get from 'lodash/get';
import { fromTo } from '@endpass/utils/date';
import { VFaucetButton } from '@endpass/faucet';
import Account from '@/components/Account';
import TokensList from '@/components/TokensList';
import { Network } from '@endpass/class';

const UPDATE_RESEND_TIMEOUT_SEC = 1000 * 60 * 2; // 2 mins

export default {
  name: 'Home',

  data() {
    return {
      isFaucetDisable: false,
      faucetTitle: 'Get test 1 ETH',
    };
  },

  computed: {
    ...mapState({
      address: state => state.accounts.address,
      activeNet: state => state.web3.activeNet,
    }),
    ...mapGetters('user', ['isLoggedIn']),
    ...mapGetters('accounts', [
      'isPublicAccount',
      'isHardwareAccount',
      'balance',
    ]),
    ...mapGetters('tokens', [
      'allCurrentAccountFullTokens',
      'currentNetUserFullTokens',
      'currentAccountFullTokens',
    ]),

    currentNetUserTokensList() {
      return Object.values(this.currentNetUserFullTokens);
    },

    currentAccountTokensList() {
      return Object.values(this.currentAccountFullTokens);
    },

    isExportable() {
      return !this.isPublicAccount && !this.isHardwareAccount;
    },

    isFaucet() {
      return this.activeNet.id === Network.NET_ID.ROPSTEN;
    },
  },

  methods: {
    onDonate() {
      this.$notify({
        title: 'Ropsten faucet ETH',
        text: 'Please wait couple of minutes for receive ETH',
        type: 'is-info',
      });

      this.$timer.start('startCountdown');
      this.faucetTitle = 'Next try after 2 mins';
      this.isFaucetDisable = true;
    },

    onDonateError(err) {
      const duration = get(err, 'response.data.duration', 0);
      const timeTitle = duration ? fromTo(0, duration) : 'in time';

      this.$notify({
        title: 'Too many attempts',
        text: `Your wallet address is banned ${timeTitle}. Please try later.`,
        type: 'is-warning',
      });

      this.faucetTitle = 'Too many attempts';
      this.isFaucetDisable = true;
    },
    onFaucetTimer() {
      this.faucetTitle = 'Get test 1 ETH';
      this.isFaucetDisable = false;
    },
  },

  mixins: [VueTimers],

  timers: {
    startCountdown: {
      repeat: false,
      time: UPDATE_RESEND_TIMEOUT_SEC,
      callback() {
        this.onFaucetTimer();
      },
    },
  },

  components: {
    Account,
    TokensList,
    VFaucetButton,
  },
};
</script>

<style lang="scss">
@media screen and (max-width: 1262px) {
  .card-content .columns {
    display: block;
  }
}
</style>
