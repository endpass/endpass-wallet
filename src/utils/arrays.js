export const uniq = arr =>
  arr.reduce((acc, item) => (acc.includes(item) ? acc : acc.concat(item)), []);

export const mapArrayByProp = (arr, prop) =>
  arr.reduce((acc, item) => {
    const target = item[prop];

    if (target) {
      return Object.assign(acc, {
        [target]: item,
      });
    }

    return acc;
  }, {});

export default {
  uniq,
  mapArrayByProp,
};
