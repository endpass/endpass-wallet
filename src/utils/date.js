import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

/**
 * Setting up dayjs globally
 */
dayjs.extend(relativeTime);

export const formateDate = date => dayjs(date).format('YYYY-MM-DD H:mm');

export const fromNow = date => dayjs(date).fromNow();

export default {
  formateDate,
  fromNow,
};
