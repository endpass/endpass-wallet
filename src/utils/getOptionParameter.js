export default function getOptionParameter(item, value) {
  return item instanceof Object ? item[value] : item;
}
