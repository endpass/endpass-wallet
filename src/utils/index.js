import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
