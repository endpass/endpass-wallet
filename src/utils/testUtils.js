import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
