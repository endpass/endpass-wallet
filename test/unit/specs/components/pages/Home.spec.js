import { shallow } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import { generateStubs } from '@/utils/testUtils';

import Home from '@/components/pages/Home';

describe('Home page', () => {
  let wrapper;
  beforeEach(() => {
    const store = new Vuex.Store({
      modules: {
        user: {
          namespaced: true,
          getters: {
            isLoggedIn: () => true,
          },
        },
        accounts: {
          namespaced: true,
          getters: {
            isPublicAccount: jest.fn(),
            isLoggedIn: jest.fn(),
            balance: jest.fn(),
          },
        },
        web3: {
          namespaced: true,
          state: {},
        },
      },
    });
    wrapper = shallow(Home, {
      store,
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('home');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
