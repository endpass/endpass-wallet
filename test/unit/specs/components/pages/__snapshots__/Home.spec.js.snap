// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page render should render initial state of the component 1`] = `
<div
  class="home-page app-page"
>
  <div>
    <div
      class="section"
    >
      <div
        class="has-text-centered"
      >
        <div
          class="card app-card main-app-card"
        >
          <div
            class="card-content"
          >
            <div>
              <h1
                class="title"
              >
                Welcome
              </h1>
               
              <p
                class="subtitle"
              >
                Get started by generating an Ethereum wallet.
              </p>
               
              <div
                class="is-centered"
              >
                <router-link
                  class="button is-success is-cta"
                  to="[object Object]"
                >
                  Create New Wallet
                </router-link>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
